src/Features/Core/Portable/Completion/Providers/ImportCompletionProvider/ExtensionMethodImportCompletionService.CacheEntry.cs

```diff
@@ -0,0 +1,151 @@
+                {
+                    _checksum = checksum;
+                    _language = langauge;
+
+                    _simpleItemBuilder = new MultiDictionary<string, DeclaredSymbolInfo>(comparer);
+                    _complexItemBuilder = ArrayBuilder<DeclaredSymbolInfo>.GetInstance();
+                }
+
+                public CacheEntry ToCacheEntry()
+                {
+                    return new CacheEntry(
+                        _checksum,
+                        _language,
+                        _simpleItemBuilder,
+                        _complexItemBuilder.ToImmutable());
+                }
+
+                public void AddItem(SyntaxTreeIndex syntaxIndex)
+                {
+                    foreach (var (targetType, symbolInfoIndices) in syntaxIndex.SimpleExtensionMethodInfo)
+                    {
+                        foreach (var index in symbolInfoIndices)
+                        {
+                            _simpleItemBuilder.Add(targetType, syntaxIndex.DeclaredSymbolInfos[index]);
+                        }
+                    }
+
+                    foreach (var index in syntaxIndex.ComplexExtensionMethodInfo)
+                    {
+                        _complexItemBuilder.Add(syntaxIndex.DeclaredSymbolInfos[index]);
+                    }
+                }
+
+                public void Dispose()
+                    => _complexItemBuilder.Free();
+            }
+        }
+
+        /// <summary>
+        /// We don't use PE cache from the service, so just pass in type `object` for PE entries.
+        /// </summary>
+        [ExportWorkspaceServiceFactory(typeof(IImportCompletionCacheService<CacheEntry, object>), ServiceLayer.Editor), Shared]
+        private sealed class CacheServiceFactory : AbstractImportCompletionCacheServiceFactory<CacheEntry, object>
+        {
+            [ImportingConstructor]
+            public CacheServiceFactory()
+            {
+            }
+        }
+
+        private static IImportCompletionCacheService<CacheEntry, object> GetCacheService(Workspace workspace)
+            => workspace.Services.GetRequiredService<IImportCompletionCacheService<CacheEntry, object>>();
+
+        private static async Task<CacheEntry?> GetCacheEntryAsync(
+            Project project,
+            bool loadOnly,
+            IImportCompletionCacheService<CacheEntry, object> cacheService,
+            CancellationToken cancellationToken)
+        {
+            // While we are caching data from SyntaxTreeInfo, all the things we cared about here are actually based on sources symbols.
+            // So using source symbol checksum would suffice.
+            var checksum = await SymbolTreeInfo.GetSourceSymbolsChecksumAsync(project, cancellationToken).ConfigureAwait(false);
+
+            // Cache miss, create all requested items.
+            if (!cacheService.ProjectItemsCache.TryGetValue(project.Id, out var cacheEntry) ||
+                cacheEntry.Checksum != checksum ||
```

Comment:
just making sure: do checksums support `==` and `!=`?

Reply:
yup
http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/Workspace/Solution/Checksum.cs,128